### ISSUE TITLE
Lazy Load Image on Flutter Web

### ISSUE THREAD

**Issue by SaltyAom** [flutter/flutter#79706]
Hi this is a feature request.

## Use case
On the web, loading a long list of the image has a critical impact on performance.

As a current state of Flutter Web, the browser load every offscreen image, increasing the loading time.

#### Left: Without lazy load, Right: With lazy load
![image](https://user-images.githubusercontent.com/35027979/113481181-45c47000-94c2-11eb-82ad-826bdde3f267.png)

Take [Flutter Gallery as an example](https://gallery.flutter.dev/#/reply), open developers tool and toggle view as iPhone 5.
An offscreen image like `avatar_4.jpg` would be immediately loading together with the onscreen image, increasing the time of loading the web.

First load, avatar_4.jpg is not a viewport and shouldn't load yet.
<img width="1440" alt="Screen Shot 2564-04-03 at 21 22 57" src="https://user-images.githubusercontent.com/35027979/113481273-c4211200-94c2-11eb-8289-a59d30823b6f.png">

After scrolling and image is in viewport, the image should load.
<img width="1440" alt="Screen Shot 2564-04-03 at 21 23 59" src="https://user-images.githubusercontent.com/35027979/113481293-ea46b200-94c2-11eb-88e1-adea0c583488.png">

Normally, a web developer can provide loading = "lazy" on an image to mark the image as lazy load.
```html
<img loading="lazy" />
```

But Flutter is rendering web in the canvas so the Flutter team might have to discuss some work about this.

Without this, a metric like Lighthouse would be recommended to [Defer offscreen images (web.dev](https://web.dev/offscreen-images).
Since Lighthouse also has a role in ranking the web on Google, web developer also usually check their web on Lighthouse as a process.

Lazy load image has a large impact on the web development side as the image is **one of the heaviest assets** on the web.
With Lazy Load Image, Flutter would have a much better performance metric.

As a current workaround of Flutter, a developer could create a lazy ListView to lazy render widget, but this also lazy render every widget, not just an image. 

On the web development side, most of the time, developers don't need to lazy render element but rather let the browser process it with HTML which has better performance than appending it using JavaScript thus having better SEO performance to let crawler read the content of the web.

This request only applies to Lazy Loading an image anywhere not inside some widget/list but rather load if it presents onscreen.

## Proposal
According to [HTML Standard (whatwg.org) or known as w3](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-loading), image tag accept `loading` attribute with a possible value of either `eager` or `lazy`.

- Where `eager` is to suddenly load the image on any condition.
```html
<img loading="eager" />
```

- And `lazy` is to load an image when the element is onscreen.
```html
<img loading="lazy" />
```

It would be a good idea to match with web development API for future usage if Flutter Web will support export as HTML and JavaScript.

Adding an optional named argument `loading` on an `Image` widget like an HTML tag, accepting an enum of `ImageLoading` (or any other name) with a default value of `ImageLoading.eager` to provide an image loading method matching property with w3 standard (currently provide `eager` and `loading`) but could also extend for Flutter's implementation way of image loading in the future.

```dart
build(context) 
  final imageLink = "https://walfiegif.files.wordpress.com/2020/12/out-transparent-23.gif?w=371&h=458"

  return Image.network(
    imageLink,
    loading: ImageLoading.lazy,
  );
}
```

<img src="https://user-images.githubusercontent.com/35027979/113482278-22042880-94c8-11eb-878a-4582b4e9114c.png" width="600" />
<img src="https://user-images.githubusercontent.com/35027979/113482280-229cbf00-94c8-11eb-95f5-bb4145e36e3c.png" width="600" />

If possible, it would be awesome to be able to use this API in native Flutter as it will reduce a lot of boilerplate in like 99% of production app with an image as lazy loading is used almost everywhere.

Opinion: This should be provided by Flutter directly as it's one of the important standards for the web optimization process.

Thanks for considering this proposal.

**Comment by iapicca** [comment]
@SaltyAom 
doesn't [PageView.builder](https://api.flutter.dev/flutter/widgets/PageView/PageView.builder.html) /[ ListView.builder](https://api.flutter.dev/flutter/widgets/ListView/ListView.builder.html) does that already?

**Comment by SaltyAom** [comment]
Correct me if I'm wrong, I'm no an expert in Flutter.
Technically, it does but not isn't the proposal

**This proposal is to request a way to only lazy load an image in a list not an entire widget in a list.**

True that PageView.builder and ListView.builder, does lazy load the image but by rendering widget in a list lazily by pagination and loads more when need but that applies to every widget as well not only the image.

Let me quote this section from above.

---
#### Current Workaround
As a current workaround of Flutter, a developer could create a lazy ListView to lazy render widget, but this also lazy render every widget, not just an image.

On the web development side, most of the time, developers don't need to lazy render element but rather let the browser process it with HTML which has better performance than appending it using JavaScript thus having better SEO performance to let crawler read the content of the web.

This request only applies to Lazy Loading an image anywhere not inside some widget/list but rather load if it presents onscreen.

---

## Usecase
Let's say that I'm building documentation in Flutter web, and I use PageView.builder or ListView.builder to the layout of main content which contains text and image, if I search to offscreen text it wouldn't be found because it hasn't rendered yet.
Although it does lazy load image it has a tradeoff where some widget which should be rendered but didn't.

Not only that I couldn't search offscreen text, but it would also have a problem with SEO and the web crawler couldn't read the collect the data because it hasn't been rendered (If Flutter Web is going to be able to export to HTML where crawler can read the web in the future).

Basically, **this proposal is to request a way to only lazy load an image in a list not an entire widget in a list.**

**Comment by iapicca** [comment]
> Not only that I couldn't search offscreen text, but it would also have a problem with SEO and the web crawler couldn't read the collect the data because it hasn't been rendered (If Flutter Web is going to be able to export to HTML where crawler can read the web in the future).

I suspect this is not how flutter works and at the moment we do have a [big seo problem](https://github.com/flutter/flutter/issues/46789) for that reason

> Basically, **this proposal is to request a way to only lazy load an image in a list not an entire widget in a list.**

"everything is a widget" - quote -

**Comment by SaltyAom** [comment]
I understand that currently, the Flutter team is working hard to solve problems with SEO, which is why I also add **if in the future after Flutter solved the problem**.

Think about this as a proposal which is not needed to be done right now but after some time in the future.
As lazy load image has a big role in web optimization, which Flutter might have to be done later to optimize the Flutter web platform.

I'm not quite sure why you quote "Everything is a Widget" on "this proposal is to request a way to only lazy load an image in a list not an entire widget in a list.".
Do you want a separated Widget for a lazy load image?

I'll provide another case so if the previous one is not clear if Flutter had solved the problem with SEO.
Take [Medium](https://medium.com) as an example.

Sometimes lazy render widget by pagination is not the best way to do lazy load image, let's say I want a web to render whole content to let web crawler read the content.
![IMG_1302](https://user-images.githubusercontent.com/35027979/113503866-b9b45600-955e-11eb-99ef-9e4de9a43f1a.jpg)
But in the content, there's also an image, which means when the web is loaded the image is also loaded.

The thing is to make whole content render except an image, to increase `First Contentful Load` which is also one of the SEO metrics by Lighthouse for ranking web on Google.

SEO is not only about the content being read by a web crawler but also a performance score provided by Lighthouse. 

Currently, Lighthouse is being used to measure the performance of the web.
The more performance the web has, the more likely it's going to have a higher rank combining with other metrics and content.

The image has a direct impact on performance measurement with Lighthouse which means it's also one of an SEO problem.

Currently, there are two workarounds with their own tradeoff.
1. Using `ListView.builder` but will sacrifice content being read from the web crawler. (If Flutter solved SEO problem)
2. Do not use any lazy render at all and let the whole content load. (This will have SEO impact on performance metric)

This proposal is to provide the third way.
**To render the contentful part and prevent the image from being loaded if it's offscreen.**

Not to lazily render widget in a list by pagination using ListView.builder.

I hope you understand, thank you.
